internal/receiver/scriptedinputsreceiver/commander.go

```diff
@@ -0,0 +1,168 @@
+// Copyright Copyright Splunk, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package scriptedinputsreceiver
+
+import (
+	"context"
+	"io"
+	"os/exec"
+	"strings"
+	"sync/atomic"
+	"syscall"
+	"time"
+
+	"go.uber.org/zap"
+)
+
+// commander can start/stop/restart the shell executable and also watch for a signal
+// for the shell process to finish.
+type commander struct {
+	name    string
+	content string
+	stdout  io.Writer
+	logger  *zap.Logger
+	cmd     *exec.Cmd
+	doneCh  chan struct{}
+	waitCh  chan struct{}
+	args    []string
+	running int64
+}
+
+func newCommander(logger *zap.Logger, name string, content string, stdout io.Writer, args ...string) *commander {
+	return &commander{
+		name:    name,
+		content: content,
+		logger:  logger,
+		args:    args,
+		stdout:  stdout,
+	}
+}
+
+// Start the shell and begin watching the process.
+// shell's stdout and stderr are written to a file.
+func (c *commander) Start(ctx context.Context) error {
+	c.logger.Info("Starting script.", zap.String("script", c.name))
+
+	c.cmd = exec.CommandContext(ctx, "sh", c.args...) //nolint:gosec
+
+	// Capture standard output and standard error.
+	c.cmd.Stdin = strings.NewReader(c.content)
+	c.cmd.Stdout = c.stdout
+	c.cmd.Stderr = c.stdout
+
+	c.doneCh = make(chan struct{}, 1)
+	c.waitCh = make(chan struct{})
+
+	if err := c.cmd.Start(); err != nil {
+		return err
+	}
+
+	c.logger.Debug("shell process started", zap.Any("PID", c.cmd.Process.Pid))
+	atomic.StoreInt64(&c.running, 1)
+
+	go c.watch()
+
+	return nil
+}
+
+func (c *commander) watch() {
+	err := c.cmd.Wait()
+	if err != nil {
+		return
+	}
+	c.doneCh <- struct{}{}
+	atomic.StoreInt64(&c.running, 0)
+	close(c.waitCh)
```

Comment:
think we'll want to defer this for error scenarios and not to block `Stop()`

Reply:
So do you want me to move these 3 lines to above if err != nil { block of code?

Reply:
this is regarding

```go
    // Wait for process to terminate
	<-c.waitCh
```

so I'd move it as to the first line in `watch()`.

```go
  defer func() { close(c.waitCh) }()
```